### DISCUSSION TITLE
Lilygo T5-Plus with micropython : getting the SD/TF card to work?

### DISCUSSION THREAD

**Discussion by brianwyld** [micropython/micropython#11460] [ESP32] [answered]
I have 2 of these boards, with ESP32-S3 WROOM-1 modules and an epaper 4.7" screen. It has a 'TF card' slot, which I would like to use from the micropython port.

From the schema, it seems like the slot is wired to use SPI and the board overview pic shows the GPIOs labelled like SPI.

So, I tried to mount the card like this:

```
from machine import SDCard, Pin
sd=SDCard()
import uos;uos.mount(sd, '/sd')
```
-> Exception:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: 16
>>>
```
Some googling led me to https://github.com/micropython/micropython/issues/4722 which implied I should try lower frequency, and also explicilty set pull_ups... 
`sd=SDCard(slot=1, width=1, sck=Pin(11,Pin.IN,Pin.PULL_UP), mosi=Pin(15,Pin.IN,Pin.PULL_UP), miso=Pin(16,Pin.IN,Pin.PULL_UP), cs=Pin(42,Pin.IN,Pin.PULL_UP), freq=400000)`
but same OSError when I try to mount it.

sd.info() gives me a timeout:
```
>>> sd.info()
sd.info()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 116] ETIMEDOUT: ESP_ERR_TIMEOUT
```

I looked at the code in machine_sdcard.c, and I see that the default 'slot' of 1 is for SD/MMC protocol, if I want to use SPi I need to set slot to 2 or 3 (for SPI2/3 on an ESP32-S3)

```
from machine import SDCard, Pin
sd=SDCard(slot=2, sck=Pin(11), mosi=Pin(15), miso=Pin(16), cs=Pin(42))
```
->exception
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: (-258, 'ESP_ERR_INVALID_ARG')
>>>
```
I can find no combination of parameters that lets me select slot 2 or 3 -> always this INVALID_ARG error.

Has anyone got the SD card on the T5-PLUS to work from micropython??

Thanks....

**Comment by lbuque** [comment]
You can refer to this [issue](https://github.com/Xinyuan-LilyGO/lilygo-micropython/issues/3)

**Reply by brianwyld** [reply]
Thanks for the pointer; however I see you used slot=2 -> this always gives me a 'ESP_ERR_INVALID_ARG' error.

**Comment by WolleKette** [comment]
A view months ago, I spent a lot of time to get SD Cards to work. Those that are present on the module as well as external ones connected via SPI. I also used an ESP32 module and I had the same strange behavior as you. Also, I could only ever mount cards once. After a umount() it didn't work anymore. I always had to do a reset. The solution was the use of the module machine.SoftSPI anstand machine.SPI. I then created a SoftSDCard class which uses the SoftSPI class. With this I can mount and unmount cards as often as I want. 

Note: I created the /sd folder in the root directory manually.

```
import os
from machine import Pin, SoftSPI
from sdcard import SDCard

class SoftSDCard:
    def __init__(self, sck=None, miso=None, mosi=None, cs=None, baudrate=100000):
        '''

        :param sck: Pin-Object or Pin-No. (int)
        :param miso: Pin-Object or Pin-No. (int)
        :param mosi: Pin-Object or Pin-No. (int)
        :param cs:  Pin-Object or Pin-No. (int)
        :param baudrate:
        '''

        self.sck = sck if isinstance(sck, Pin) else Pin(sck)
        self.miso = miso if isinstance(miso, Pin) else Pin(miso)
        self.mosi = mosi if isinstance(mosi, Pin) else Pin(mosi)
        self.cs = cs if isinstance(cs, Pin) else Pin(cs)
        self.baudrate = baudrate

        self.__spi = None
        self.__sd = None

    def mount(self):
        try:
            if self.__spi == None:
                self.__spi = SoftSPI(baudrate=100000, sck=self.sck, miso=self.miso, mosi=self.mosi)
    
            self.__sd =  SDCard(self.__spi, self.cs)

            os.mount(self.__sd, '/sd')

            return True
        except:
            return False

    def umount(self):
        try:
            os.umount('/sd')
            self.__sd = None
        except:
            pass

    def get_content(self):
        try:
            content = os.listdir('/sd')
            return content
        except:
            return None

```

**Reply by brianwyld** [reply]
Thanks, I'll give it a go with SoftSPI!

**Reply by WolleKette** [reply]
Please post if it worked.

**Comment by brianwyld** [comment]
Finally back to this project... but still stuck on this:
```
# SD card with softspi
import os
from machine import Pin, SoftSPI
spi = SoftSPI(baudrate=100000, sck=Pin(11,Pin.IN,Pin.PULL_UP), miso=Pin(16,Pin.IN,Pin.PULL_UP), mosi=Pin(15,Pin.IN,Pin.PULL_UP))
from sdcard import SDCard
```
spi creation ok, but then fails as no 'sdcard' module in the build...

-> try with machine.SDCard
```
from machine import SDCard)
sd=SDCard(spi=spi, cs=Pin(42,Pin.IN,Pin.PULL_UP))
```
No, spi not an  argument for the builtin machine.SDCard

Need to search for the sdcard module that can take a spi instance...

**Comment by brianwyld** [comment, marked as answer]
By copying in the sdcard.py from micropython/micropython-lib/micropython/drivers/storage/sdcard.py, and using the SDCard class from there, I finally got it to work with SoftSPI:
```
from machine import Pin, SoftSPI
from sdcard import SDCard
import uos
spi = SoftSPI(baudrate=100000, sck=Pin(11,Pin.IN,Pin.PULL_UP), miso=Pin(16,Pin.IN,Pin.PULL_UP), mosi=Pin(15,Pin.IN,Pin.PULL_UP))
sd=SDCard(spi=spi, cs=Pin(42,Pin.IN,Pin.PULL_UP))
uos.mount(sd, '/sd')
uos.listdir("/sd")
```
It also works with machine.SPI using SPI instance 2:
spi = SPI(2, baudrate=400000, sck=Pin(11,Pin.IN,Pin.PULL_UP), miso=Pin(16,Pin.IN,Pin.PULL_UP), mosi=Pin(15,Pin.IN,Pin.PULL_UP))
but the first attempt at mount gave me:
```
>>> uos.mount(sd, '/sd')
uos.mount(sd, '/sd')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/flash/sdcard.py", line 258, in readblocks
  File "/flash/sdcard.py", line 196, in readinto
OSError: timeout waiting for response
```
Doing the mount again, it was ok... Not sure why....

**Comment by lbuque** [comment]
It works.

The firmware I am using is https://github.com/lbuque/micropython-builder.

You can find the compiled firmware in the github action.

```python
import machine
import os

sd = machine.SDCard(slot=3, sck=11, miso=16, mosi=15, cs=42)
sd.info()
os.mount(sd, "/sd")
os.listdir('/sd')
```

![image](https://github.com/micropython/micropython/assets/46518962/3af05981-50bf-4603-9b7a-5b86df8b11b0)

**Reply by lbuque** [reply]
When slot=2, it cannot work. The cause is unknown.

**Reply by brianwyld** [reply]
my micropython project is also a fork of the same version, but when I try the same as you I get:
```
>>> sd = machine.SDCard(slot=3, sck=11, miso=16, mosi=15, cs=42)
sd = machine.SDCard(slot=3, sck=11, miso=16, mosi=15, cs=42)
E (30970) spi: spi_bus_initialize(762): invalid dma channel, chip only support spi dma channel auto-alloc
E (30970) sdspi_host: spi_bus_initialize failed with rc=0x102
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: (-258, 'ESP_ERR_INVALID_ARG')
```
Same thing with SPI channel 2...

But if I use the machine.SPI on channel 2, with the SDCard class from the micropython-lib project, all is ok:
```
>>> from machine import Pin, SPI
from machine import Pin, SPI
>>> spi = SPI(2, baudrate=400000, sck=Pin(11,Pin.IN,Pin.PULL_UP), miso=Pin(16,Pin.IN,Pin.PULL_UP), mosi=Pin(15,Pin.IN,Pin.PULL_UP))
spi = SPI(2, baudrate=400000, sck=Pin(11,Pin.IN,Pin.PULL_UP), miso=Pin(16,Pin.IN,Pin.PULL_UP), mosi=Pin(15,Pin.IN,Pin.PULL_UP))
>>> from sdcard import SDCard
from sdcard import SDCard
>>> sd=SDCard(spi=spi, cs=Pin(42,Pin.IN,Pin.PULL_UP))
sd=SDCard(spi=spi, cs=Pin(42,Pin.IN,Pin.PULL_UP))
>>> sd.info()
sd.info()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'SDCard' object has no attribute 'info'
>>> import uos
import uos
>>> uos.mount(sd, '/sd')
uos.mount(sd, '/sd')
>>> uos.listdir("/sd")
uos.listdir("/sd")
['System Volume Information', 'imgs', 'fonts']
```
note this does not work with SPI channel 3 : tells me it doesn't exist!

Just to confirm, you are using a T5-PLUS board (with the ESP32-S3 chip)?

**Reply by lbuque** [reply]
Yes, I use T5-4.7 PLUS.
